// react-formik-step/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var formik = require('formik');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

var FormContext = React.createContext({
    steps: [],
    currentStepIndex: 0,
    gotoStep: function () { }
});
FormContext.displayName = "StepForm";
var useStepContext = function () {
    return React.useContext(FormContext);
};
function StepForm(_a) {
    var onSubmit = _a.onSubmit, children = _a.children, formikProps = __rest(_a, ["onSubmit", "children"]);
    var _b = React.useState(0), step = _b[0], setStep = _b[1];
    var formSteps = React.Children.toArray(children);
    var activeStep = formSteps[step];
    var gotoStep = function (step) {
        var normalized = Math.min(step, formSteps.length - 1);
        normalized = Math.max(step, 0);
        setStep(normalized);
    };
    var metas = formSteps.map(function (s) { return s.props; });
    var state = {
        steps: metas,
        currentStepIndex: step,
        gotoStep: gotoStep
    };
    var stepSubmit = function (values, helpers) {
        var isLastStep = step === formSteps.length - 1;
        if (isLastStep) {
            onSubmit(values, helpers);
        }
        else {
            helpers.setTouched({});
            helpers.setSubmitting(false);
            gotoStep(step + 1);
        }
    };
    var validate = function (values) {
        return activeStep.props.validate ? activeStep.props.validate(values) : {};
    };
    return (React__default.createElement(formik.Formik, __assign({}, formikProps, { onSubmit: stepSubmit, validate: validate, validationSchema: activeStep.props.validationSchema }), function (_a) {
        var handleSubmit = _a.handleSubmit;
        return (React__default.createElement("form", { onSubmit: handleSubmit },
            React__default.createElement(FormContext.Provider, { value: state }, activeStep)));
    }));
}

var Step = function (_a) {
    var children = _a.children;
    return React__default.createElement(React__default.Fragment, null, children);
};

exports.Step = Step;
exports.StepForm = StepForm;
exports.useStepContext = useStepContext;
